Allow rails 5.2, update travis build matrix

lostapathy opened the pull request:
This loosens up the gemspec to allow it to be used with rails 5.2 apps.

I also setup Appraisal to manage Gemfiles to test against various rails versions and updated the Travis build matrix to match.

If maintainers are open to removing all unsupported ruby/rails version from the build matrix, I'd like to do that but I didn't want to remove anything from the test matrix.

lostapathy commented:
I think this should be good to go now, but Travis seems to be having network issues setting up a couple of the builds.

I went back to full ruby versions in the travis config, as it was trying to build against really old rubies when configured with the short aliases.

The second commit to `gem update --system` can probably come back out in the future, but it's needed now, see https://github.com/travis-ci/travis-ci/issues/8978

Thanks for a cool gem!

lostapathy commented:
This is ready for review.  Travis have fixed their issue and this build is now green.  Thanks!

varyonic commented on has_scope.gemspec:
```diff
@@ -24,6 +24,9 @@ Gem::Specification.new do |s|
 
-  s.add_runtime_dependency 'actionpack', '>= 4.1', '< 5.2'
-  s.add_runtime_dependency 'activesupport', '>= 4.1', '< 5.2'
+  s.add_runtime_dependency 'actionpack', '>= 4.1', '< 5.3'
```
As a nice to have, I think '<= 5.2.x' works with 5.2.beta, which is a bit more elegant, as the next release is probably 6.0.

varyonic commented on .travis.yml:
```diff
@@ -1,12 +1,32 @@
+  - 2.2.9
+  - 2.3.6
+  - 2.4.3
+  - 2.5.0
```
Is there a reason for specifying patch levels?

lostapathy replied:
It seems to vary over time - it's not totally stable over time and leads to non-obvious build failures when the problem comes back after a while.  I've seen several projects go to this method to avoid that, Rails being an obvious example.

When I did the first commit of this patch, 2 or 3 of them were out of date so I made this change and force-pushed over the top.

varyonic commented on gemfiles/rails_4.1.gemfile:
```diff
@@ -0,0 +1,8 @@
+# This file was generated by Appraisal
```
These are simple gemfiles: could they be maintained by hand and Appraisal dispensed with?

lostapathy replied:
Yes, but, Appraisal also also gives you the ability to (for example) `bundle exec appraisal rake test` to run the test suite against all gemfiles at once.  That's super helpful when you're trying to keep a library working against a range of dependency versions, as is the case here.

It's certainly a "nice to have" and I can back that out if that's what it takes to get this merged.  I'm not aware of any projects adding Appraisal then regretting it and backing it out, FWIW.

varyonic replied:
https://github.com/activeadmin/activeadmin/commit/b3d3ccb4232eb9edc99740ddc4d875482d93dc4e FWIW. That was @deivid-rodriguez's call, here I think it is up to @rafaelfranca

deivid-rodriguez replied:
Yup, I don't think `appraisal` provides a lot of benefit.

If a task like `bundle exec appraisal rake test` was to be run frequently by developers, it'd be a matter of adding a simple `Rakefile` task that iterates over the available gemfiles and runs tests on each using the proper `BUNDLE_GEMFILE` environment.

lostapathy replied:
I don't want to call out other projects by name for making a mess, but I've worked on gems that started with a "simple rake task" and grew it into a collection of fragile, unmaintanable rake tasks as they reinvented all the Appraisal functionality.

deivid-rodriguez replied:
Each to their own, I guess :)

varyonic commented on Appraisals:
```diff
@@ -0,0 +1,24 @@
+
+appraise 'rails-5.2' do
+  gem 'actionpack', '~> 5.2.0.beta2'
+  gem 'activesupport', '~> 5.2.0.beta2'
```
'~> 5.2.x'

lostapathy replied:
Thanks! Did not realize you could specify this way and pick up the beta.

varyonic commented:
@rafaelfranca ?

varyonic commented:
Closes #92

gobijan commented:
I cloesed my PR since this one is already pending.

rafaelfranca commented on .gitignore:
```diff
@@ -1,2 +1,3 @@
 .bundle
+Gemfile.lock
```
Please add back the Gemfile.lock.

rafaelfranca commented on Appraisals:
```diff
@@ -0,0 +1,24 @@
+appraise 'rails-4.1' do
```
We don't need appraisal to generate gemfiles.

rafaelfranca commented on has_scope.gemspec:
```diff
@@ -24,6 +24,9 @@ Gem::Specification.new do |s|
 
-  s.add_runtime_dependency 'actionpack', '>= 4.1', '< 5.2'
-  s.add_runtime_dependency 'activesupport', '>= 4.1', '< 5.2'
+  s.add_runtime_dependency 'actionpack', '>= 4.1', '<= 5.2'
```
I think here we want `< 6.0`

clairity commented:
now that rails 5.2 is released, is there any chance we can get this merged soon? the `'< 5.2'` actionpack stipulation (at least) is blocking the update to 5.2 (through a dependency of activeadmin)?

lostapathy commented:
@rafaelfranca this is updated as requested.

rafaelfranca commented:
I'll take care of it, thanks

lostapathy commented:
@rafaelfranca I removed it from the gemspec and removed the file, all that's left is the comment in the files it generated.  Do you want me to remove that and squash it all?

varyonic commented:
@rafaelfranca @tegon I just redid this without Appraisal for you a couple of hours ago, see above.
